### PR TITLE
[state_dict] Add cpu_only and ranks_only support for _gather_state_dict

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112837
* #112885
* __->__ #112836

Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

### CODE REVIEW

File: torch/distributed/checkpoint/_state_dict_utils.py [hunk 1]
```diff
@@ -49,13 +49,34 @@ def _all_gather_sharded_tensor(
...
     """
-    Given a state_dict, this API gathers all the ShardedTensors or DTensors in the state_dict.
+    Given a state_dict, this API gathers all the ShardedTensors or DTensors in
+    the state_dict.
+
+
+    Args:
+        state_dict (Dict[str, Any]): the target sharded state_dict.
+        pg (Optional[dist.ProcessGroup]): the process group that is used to
+            gather ShardedTensor.
```
Comment (wz337) [inline torch/distributed/checkpoint/_state_dict_utils.py:66]: nit: gather DTensor or ShardedTensor.

File: torch/distributed/checkpoint/_state_dict_utils.py [hunk 2]
```diff
@@ -49,13 +49,34 @@ def _all_gather_sharded_tensor(
...
+        device: (Optional[torch.device]): the device that is used to
+            perform allgather for ShardedTensor.
```
Comment (wz337) [inline torch/distributed/checkpoint/_state_dict_utils.py:68]: nit: gather DTensor or ShardedTensor.

File: torch/distributed/checkpoint/_state_dict_utils.py [hunk 3]
```diff
@@ -86,7 +107,11 @@ def _gather_state_dict(
...
             value = value.to_local()
         elif isinstance(value, dict):
-            value = _gather_state_dict(value, pg, device)
+            value = _gather_state_dict(value, pg=pg, device=device)
+
+        if isinstance(value, torch.Tensor) and cpu_offload:
+            value = value.to(cpu_device)
 
-        new_state_dict[key] = value
+        if not cpu_offload or len(ranks_only) == 0 or dist.get_rank(pg) in ranks_only:
```
Comment (wz337) [inline torch/distributed/checkpoint/_state_dict_utils.py:115]: Is `dist.get_rank(pg) in ranks_only` along not sufficient?

Comment (fegin) [reply torch/distributed/checkpoint/_state_dict_utils.py:115]: ranks_only can be empty (the default value). But it seems that we should not restrict this function to be cpu_offload only.

File: torch/distributed/checkpoint/_state_dict_utils.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/checkpoint/_state_dict_utils.py:68]: ShardedTensor only. DTensor uses DeviceMesh to gather the tensors.

**Review by LucasLLC — APPROVED** [review]
lgtm!

File: torch/distributed/checkpoint/_state_dict_utils.py (hunk 3 above)
Comment (wz337) [reply torch/distributed/checkpoint/_state_dict_utils.py]: I see. Makes sense!

**Review by wz337 — APPROVED** [review]

**Pull request closed without merging**